packages/client-node/src/connection/node_base_connection.ts

```diff
@@ -568,79 +581,94 @@ export abstract class NodeBaseConnection
+                  this.logger.trace({
+                    message: `Removing socket ${socketId} after ${this.idleSocketTTL} ms of idle`,
+                  })
+                  this.knownSockets.delete(socket)
+                  socket.destroy()
+                }, this.idleSocketTTL).unref()
+                this.knownSockets.set(socket, {
+                  id: socketId,
+                  idle_timeout_handle: idleTimeoutHandle,
+                })
               })
-            })
 
-            const cleanup = () => {
-              const maybeSocketInfo = this.knownSockets.get(socket)
-              // clean up a possibly dangling idle timeout handle (preventing leaks)
-              if (maybeSocketInfo?.idle_timeout_handle) {
-                clearTimeout(maybeSocketInfo.idle_timeout_handle)
+              const cleanup = () => {
+                const maybeSocketInfo = this.knownSockets.get(socket)
+                // clean up a possibly dangling idle timeout handle (preventing leaks)
+                if (maybeSocketInfo?.idle_timeout_handle) {
+                  clearTimeout(maybeSocketInfo.idle_timeout_handle)
+                }
+                this.logger.trace({
+                  message: `Socket ${socketId} was closed or ended, 'free' listener removed`,
+                })
               }
+              socket.once('end', cleanup)
+              socket.once('close', cleanup)
+            } else {
+              clearTimeout(socketInfo.idle_timeout_handle)
               this.logger.trace({
-                message: `Socket ${socketId} was closed or ended, 'free' listener removed`,
+                message: `Reusing socket ${socketInfo.id}`,
+              })
+              this.knownSockets.set(socket, {
+                ...socketInfo,
+                idle_timeout_handle: undefined,
               })
             }
-            socket.once('end', cleanup)
-            socket.once('close', cleanup)
-          } else {
-            clearTimeout(socketInfo.idle_timeout_handle)
-            this.logger.trace({
-              message: `Reusing socket ${socketInfo.id}`,
-            })
-            this.knownSockets.set(socket, {
-              ...socketInfo,
-              idle_timeout_handle: undefined,
-            })
           }
+        } catch (e) {
+          logger.error({
+            message: 'An error occurred while housekeeping the idle sockets',
+            err: e as Error,
+          })
         }
 
         // Socket is "prepared" with idle handlers, continue with our request
         pipeStream()
 
         // This is for request timeout only. Surprisingly, it is not always enough to set in the HTTP request.
-        // The socket won't be actually destroyed, and it will be returned to the pool.
+        // The socket won't be destroyed, and it will be returned to the pool.
         socket.setTimeout(this.params.request_timeout, onTimeout)
       }
 
       function onTimeout(): void {
```

Comment:
it's annoying to add manually to every event handler. I wish we had a better option  😞